runtime/src/vote_account.rs

```diff
@@ -119,9 +125,9 @@ impl VoteAccounts {
     fn add_node_stake(&mut self, stake: u64, vote_account: &ArcVoteAccount) {
         if stake != 0 && self.staked_nodes_once.is_completed() {
             if let Some(node_pubkey) = vote_account.node_pubkey() {
-                self.staked_nodes
-                    .write()
-                    .unwrap()
+                let mut staked_nodes = self.staked_nodes.write().unwrap();
+                let staked_nodes = Arc::make_mut(&mut staked_nodes);
+                staked_nodes
```

Comment:
hmmmm does this only add the new entry to the local copy of `staked_nodes`? This won't persist the insert into the `self.staked_nodes` for other callers to observe right?

From what I can tell this function is only called from `Bank::store_account()` which is used in ledger-tool and probably expects the store to be persisted.

Reply:
It clones the inner pointer, so the changes are persisted.
Please see the added `test_staked_nodes_cow` or the example in `Arc::make_mut`:
https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut

Reply:
I see, I mixed up what was being cloned, thanks for the clarification.

And the cost of cloning is acceptable because `Bank::store_account()` is only used in ledger-tool where there are no other outstanding references that would trigger this clone?

To avoid silent perf issues in the future if somebody calls `store_account` more intensively,  should we be more explicit about this and use `get_mut()` instead of `make_mut()` and return a boolean from this function to indicate success/failure,

Reply:
This code only makes a clone if:
* `.staked_nodes` was invoked since last mutation (i.e. since last `make_mut`).
* That returned `Arc<...>` is still alive when you want to mutate the hash-map again (i.e. when you call `make_mut`).

Current master code is making a clone every time `.staked_ndoes` is invoked:
https://github.com/solana-labs/solana/blob/fd937548a/runtime/src/vote_account.rs#L77

So this code always does fewer clones than current master code, and `make_mut` does not necessarily make a clone.

Reply:
> That returned Arc<...> is still alive when you want to mutate the hash-map again (i.e. when you call make_mut).

Right, I'm worried about these calls https://github.com/solana-labs/solana/blob/fd937548a01ecbfae4fd0fef25efe4ad029dcffe/ledger-tool/src/main.rs#L2064-L2067 silently incurring a clone in the future if some other thread is holding another reference to this `Arc`. I feel like being explicit with a result from something like a `store_without_clone` that we can assert on makes this more transparent, and avoids silently incurring this cost.

Reply:
Can you please describe the scenario that previously it was not making a clone and now it does?

1. Previously [*every time* one calls `.staked_nodes()` it had to make a clone](https://github.com/solana-labs/solana/blob/fd937548a/runtime/src/vote_account.rs#L77). No exceptions there.
2. Now it clones only if *all* these happen:
    * you call `.staked_nodes()`, 
    * and hold on to the returned `Arc<...>` 
    * and then also try to mutate stake-nodes while holding to that `Arc`.

(2) is a subset of (1), so for every time that this code ends up making a clone, it was making a clone previously anyways; and so this code is removing clones, not adding clones.

Reply:
>Can you please describe the scenario that previously it was not making a clone and now it does?

I don't think there is one currently, was referring to future-proofing calls like https://github.com/solana-labs/solana/blob/fd937548a01ecbfae4fd0fef25efe4ad029dcffe/ledger-tool/src/main.rs#L2064-L2067 in case anybody does add a case where they hold a returned `Arc`. This way anybody calling these functions will be cognizant of when a clone is incurred.

Reply:
Still like current call-sites, this code cannot introduce any *new* clones (where with the old code there was no clone).
I would be happy to discuss this more over a call or something next time we chat. For now, any objections to merge this?

Reply:
> I would be happy to discuss this more over a call or something next time we chat
Sounds good, no other objections!

Reply:
thank you :pray: